src/Common/HashTable/HashTableKeyHolder.h

```diff
@@ -132,4 +135,103 @@ inline void ALWAYS_INLINE keyHolderDiscardKey(DB::SerializedKeyHolder & holder)
     holder.key.data = nullptr;
     holder.key.size = 0;
 }
+namespace DB
+{
+struct AdaptiveKeysHolder
```

Comment:
could you pls provide a brief description of the introduced data structures: what purpose do they serve, how they're connected. anything that could facilitate review